### PR TITLE
os/bluestore: make deferred writes less aggressive for large writes

### PR DESCRIPTION
**Pull request by ifed01**
Fixes: https://tracker.ceph.com/issues/52089

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -14241,14 +14261,19 @@ int BlueStore::_do_alloc_write(
 
     PExtentVector extents;
     int64_t left = final_length;
+    bool has_chunk2defer = false;
     while (left > 0) {
       ceph_assert(prealloc_left > 0);
+      has_chunk2defer |= (prealloc_pos_length < prefer_deferred_size.load());
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14267]: s/prefer_deferred_size.load()/prefer_deferred_size_snapshot
I did not notice that when I wrote this code.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -14241,14 +14261,19 @@ int BlueStore::_do_alloc_write(
 
     PExtentVector extents;
     int64_t left = final_length;
+    bool has_chunk2defer = false;
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14264]: +1 Good name.

File: src/test/objectstore/store_test.cc [hunk 3]
```diff
@@ -7483,7 +7630,8 @@ TEST_P(StoreTestSpecificAUSize, DeferredDifferentChunks) {
 	      CEPH_OSD_OP_FLAG_FADVISE_NOCACHE);
       r = queue_transaction(store, ch, std::move(t));
       ++exp_bluestore_write_big;
-      ++exp_bluestore_write_big_deferred;
+      if (expected_write_size != prefer_deferred_size)
```
Comment (aclamk) [inline src/test/objectstore/store_test.cc:7633]:
```
if (expected_write_size > prefer_deferred_size)
```
?

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -13927,13 +13939,17 @@ void BlueStore::_do_write_big(
             _do_write_big_apply_deferred(txc, c, o, tail_info,
               blp, wctx);
           }
+	  dout(20) << __func__ << " defer big: 0x" << std::hex
+		   << offset << "~" << l
+		   << std::dec << dendl;
           offset += l;
           length -= l;
-          logger->inc(l_bluestore_write_big_blobs, remaining ? 2 : 1);
+	  logger->inc(l_bluestore_write_big_blobs, remaining ? 2 : 1);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:13947]: line touched, but unchanged.

**Review by aclamk — APPROVED** [review]
Good stuff.

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -13845,14 +13849,22 @@ void BlueStore::_do_write_big(
   uint64_t prefer_deferred_size_snapshot = prefer_deferred_size.load();
   while (length > 0) {
     bool new_blob = false;
-    uint32_t l = std::min(max_bsize, length);
+    uint32_t l = p2nphase(offset, max_bsize);
+    if (!l) {
+      l = max_bsize;
+    }
+    l = std::min(uint64_t(l), length);
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: replaced with 
      l = max_bsize - p2phase(offset, max_bsize);
      l = std::min(uint64_t(l), length);

File: src/test/objectstore/store_test.cc (hunk 3 above)
Comment (ifed01) [reply src/test/objectstore/store_test.cc]: fixed

File: src/os/bluestore/BlueStore.cc (hunk 4 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: fixed

**Comment by neha-ojha** [comment]
started a rados run with latest changes included: https://pulpito.ceph.com/nojha-2021-08-11_00:04:11-rados-wip-42725-2021-08-10-distro-basic-smithi/

**Comment by markhpc** [comment]
I tested out a couple of simple code changes in my wip branch here (fix1-5 change different aspects of the deferred IO path):

https://github.com/markhpc/ceph/tree/wip-bs-deferred-fix

Adam created a PR with a different approach at a fix here:

https://github.com/ceph/ceph/pull/42721

And Igor's PR here that incorporates Adam's fix and goes much farther changing the deferred IO path.

I tested all of these using the same benchmark configuration I was testing previously, but also in more recent tests reverted a recent PR causing a performance regression in master (Thanks to @mkogan1 for finding this).  There appears to be  *another* performance regression in master I haven't tracked down, but I will try to do so this week.

https://docs.google.com/spreadsheets/d/1mTnKvLh8NIxPukad0o0f8ao-T2-GORXWASTQBdxtDFM/edit?usp=sharing

As a reminder, these tests are looking at deferred IO behavior on NVMe drives rather than HDDs, so the performance numbers are not necessarily relevant (but may be depending on the situation).  The more important numbers to look at are the memtable flushing and compaction behavior (which one of our teams at Red Hat did observe as excessively high in pacific vs nautilus based builds on HDD).  Generally speaking it looks like #42725 is behaving pretty similarly to FIX1/3/5 in my tests where:

https://github.com/markhpc/ceph/commit/b0a0482d92ee5387298e34aa6ebd2d40a634713a#diff-6f3b1a5dcf5313dc06f71ba4997ee28ec219ecb5633a728b916465c684f1d87dL14298-R14299

What's strangely surprising is that there seems to be some yet-unknown affect on the 128KB random read path that is causing a bimodal performance distribution.  It seems like the deferred IO path may be having an affect, but it's not consistent (for instance deferred=0k in pacific was bad, deferred=0k in master was good, deferred=64k in pacific was good, but deferred=64k master was bad)

So far however, it's debatable if any of these fixes are quite as good as just increasing the blob size in pacific to 256K (though I haven't tested that configuration in master yet).  It's possible that Igor's PR here may confer some advantages for small writes (they are much faster with this PR), but I need to verify that we are actually properly deferring those writes in this PR and not just behaving like prefer_deferred=0k.

**Pull request merged**